linetools/analysis/cog.py

```diff
@@ -0,0 +1,152 @@
+"""
+Module on Curve of Growth
+"""
+from __future__ import print_function, absolute_import, division, unicode_literals
+
+import numpy as np
+import sys
+import os
+import warnings
+import pdb
+
+from scipy import integrate
+from scipy.interpolate import interp1d
+
+from astropy import units as u
+from astropy.modeling import FittableModel, Parameter
+from astropy.modeling import fitting
+
+#from xastropy.xutils import xdebug as xdb
+
+# Begin
+def ftau_intgrnd(x,tau0=0.1):
+    return 1 - np.exp(-tau0 * np.exp(-x**2))
+
+# Generate Ftau (could archive, but this is reasonably fast)
+neval = 10000
+lgt = np.linspace(-3, 9, neval)
+all_tau0 = 10.**lgt
+#
+xFtau0 = np.zeros(neval)
+for jj,tau0 in enumerate(all_tau0):
+    xFtau0[jj], ferr = integrate.quad(ftau_intgrnd, 0, np.inf, args=(tau0,))
+
+# Now interpolate
+intFtau0 = interp1d(all_tau0, xFtau0, bounds_error=False,fill_value=0.)
+
+##############################
+def cog_plot(COG_dict):
+    """Generate a plot for COG solution
+    Parameters
+    ----------
+    COG_dict : dict
+      dict containing the COG inputs and solution
+    """
+    import matplotlib.pyplot as plt
+    import matplotlib as mpl
+    mpl.rcParams['font.family'] = 'stixgeneral'
+    mpl.rcParams['font.size'] = 17.
+    # Plot
+    plt.clf()
+    ax = plt.gca()
+    # Data with values
+    gdv = COG_dict['redEW'] > 0.
+    yerr=(COG_dict['sigEW'][gdv]/COG_dict['wrest'][gdv])/COG_dict['redEW'][gdv]
+    ax.errorbar(np.log10(COG_dict['f'][gdv]*COG_dict['wrest'][gdv].to('cm').value),
+                np.log10(COG_dict['redEW'][gdv]),
+                yerr=yerr, fmt='o')
+    # Upper limit
+    upper = COG_dict['redEW'] <= 0.
+    ax.scatter(np.log10(COG_dict['f'][upper]*COG_dict['wrest'][upper].to('cm').value),
+               np.log10(3*COG_dict['sigEW'][upper]/COG_dict['wrest'][upper]), color='red', marker='v')
```

Comment:
This is fine for upper limits. But you can also add them using plt.errorbar and they look nicer: http://matplotlib.org/examples/statistics/errorbar_limits.html